src/linear_algebra/clifford_algebra/contraction.lean

```diff
@@ -0,0 +1,313 @@
+import linear_algebra.clifford_algebra.fold
+import linear_algebra.clifford_algebra.grading
+import linear_algebra.clifford_algebra.conjugation
+
+/-!
+# Contraction in Clifford Algebras
+
+This file contains some of the results from [grinberg_clifford_2016][].
+The key result is `clifford_algebra.equiv_exterior`.
+
+## Main definitions
+
+* `clifford_algebra.apply_dual_left`: contract a multivector by a `module.dual R M` on the left.
+* `clifford_algebra.apply_dual_right`: contract a multivector by a `module.dual R M` on the right.
+* `clifford_algebra.alpha`: convert between two algebras of different quadratic form, sending
+  vectors to vectors. The difference of the quadratic forms must be a bilinear form.
+* `clifford_algebra.equiv_exterior`: in characteristic not-two, the `clifford_algebra Q` is
+  isomorphic as a module to the exterior algebra.
+
+## Implementation notes
+
+This file somewhat follows [grinberg_clifford_2016][], although we are missing some of the induction
+principles needed to prove many of the results. Here, we avoid the quotient-based approach described
+in [grinberg_clifford_2016][], instead directly constructing our objects using the universal
+property.
+
+Note that [grinberg_clifford_2016][] concludes that its contents are not novel, and are in fact just
+a rehash of parts of Bourbaki; we should at some point consider swapping our references to refer to
+the latter.
+
+Within this file, we use the local notation
+* `x ⌊ d` for `apply_dual_right x d`
+* `d ⌋ x` for `apply_dual_left d x`
+
+-/
+
+universes u1 u2 u3
+
+variables {R : Type u1} [comm_ring R]
+variables {M : Type u2} [add_comm_group M] [module R M]
+variables (Q : quadratic_form R M)
+
+namespace clifford_algebra
+
+section apply_dual_left
+
+variables (d d' : module.dual R M)
+
+/-- Auxiliary construction for `clifford_algebra.apply_dual_left` -/
+@[simps]
+def apply_dual_left_aux (d : module.dual R M) :
+  M →ₗ[R] clifford_algebra Q × clifford_algebra Q →ₗ[R] clifford_algebra Q :=
+begin
+  have v_mul := (algebra.lmul R (clifford_algebra Q)).to_linear_map ∘ₗ (ι Q),
+  exact d.smul_right (linear_map.fst _ (clifford_algebra Q) (clifford_algebra Q)) -
+        v_mul.compl₂ (linear_map.snd _ (clifford_algebra Q) _),
+end
+
+lemma apply_dual_left_aux_apply_dual_left_aux (v : M) (x : clifford_algebra Q)
+  (fx : clifford_algebra Q) :
+  apply_dual_left_aux Q d v (ι Q v * x, apply_dual_left_aux Q d v (x, fx)) = Q v • fx :=
+begin
+  simp only [apply_dual_left_aux_apply_apply],
+  rw [mul_sub, ←mul_assoc, ι_sq_scalar, ←algebra.smul_def, ←sub_add, mul_smul_comm, sub_self,
+    zero_add],
+end
+
+variables {Q}
+
+/-- Contract an element of the exterior algebra with an element `d : module.dual R M` from the left.
+
+Note that $v ⌋ x$ is spelt `apply_dual_left (Q.associated v) x`.
+
+This includes [grinberg_clifford_2016][] Theorem 10.75 -/
+def apply_dual_left : module.dual R M →ₗ[R] clifford_algebra Q →ₗ[R] clifford_algebra Q :=
```

Comment:
Should these names use `contract` instead of `apply`?

Reply:
That's probably a reasonable suggestion. I only use the word `apply` because mathlib represents `module.dual` as a function.